### PR TITLE
Menu performance

### PR DESCRIPTION
**Pull request by laurafeier**

### CODE REVIEW

File: cms/menu.py [hunk 1]
```diff
@@ -17,6 +18,39 @@
...
+
+def _pages_to_dict(pages):
+    """
+        Builds a dict with page ids as keys and list of children ids as values
+    """
+    # make sure all pages are present as keys
+    pages_with_children = defaultdict(list)
+    for page in pages:
+        # set page as key
+        pages_with_children[page.id]
```
Comment (kux) [inline cms/menu.py]: I know this creates a new entry with an empty list, but it still looks very strange. Make it explicit:

```
pages_with_children[page.id] = []
```

(ps: the zen of python rule2: Explicit is better than implicit. :) )

File: cms/menu.py [hunk 2]
```diff
@@ -17,6 +18,39 @@
...
+
+def _get_descendants(pages_dict, page_ids):
+    """
+        Returnes all descendants for a list of pages using a page dict
+            built with _pages_to_dict method.
+    """
+    result = []
+    for page_id in page_ids:
+        result.append(page_id)
+        value = pages_dict.get(page_id, None)
```
Comment (kux) [inline cms/menu.py]: I think `children_ids` would sound better than `value`.

File: cms/menu.py [hunk 3]
```diff
@@ -17,6 +18,39 @@
...
+    """
+    # make sure all pages are present as keys
+    pages_with_children = defaultdict(list)
+    for page in pages:
+        # set page as key
+        pages_with_children[page.id]
+        # if page has parent add page to parent's list of children
+        parent_id = page.parent_id
+        # parent might not be published
+        if parent_id and parent_id in pages_with_children:
```
Comment (kux) [inline cms/menu.py]: Shouldn't it be:

```
if parent_id and parent_id in pages:
```

If pages = [1, 3, 5] and page 3 is the parent of page 1, the dict will have an empty list for page 3.

Comment (laurafeier) [inline cms/menu.py]: True.

File: cms/menu.py [hunk 4]
```diff
@@ -28,104 +62,118 @@ def get_visible_pages(request, pages, site=None):
...
+
         #no page perms edgcase - all visible
         if ((is_setting_public_all or (
-            is_setting_public_staff and request.user.is_staff))and 
-            not restricted_pages and
+            is_setting_public_staff and request.user.is_staff))and
+            not restricted and
             not global_view_perms):
-            return [page.pk for page in pages]
+            return pages_with_children.keys()
```
Comment (kux) [inline cms/menu.py]: I don't like the idea of 

```
return pages_with_children.keys()
```

The reason for this is the fact that pages_with_children is a defaultdict, which means it's quite easy for some future developer to do a `pages_with_children[some value]` ignoring the fact that `some value` wasn't in the initial pages list. At the end of the day, being able to do this is the whole point of having a default dict.

I actually wanted to suggest replacing

```
        value = pages_dict.get(page_id, None)
        if value:
            result.extend(_get_descendants(pages_dict, value))
```

with

```
        result.extend(_get_descenants(page_dict, page_dict[page_id]))
```

but then realized the potential side effects.

Comment (kux) [inline cms/menu.py]: Therefore, I would suggest saving the initial page pks in a separate list and probably also add assertions like
`assert set(page_with_children.keys()) == set(initial_pks)` just to make sure that developers don't accidentally add stuff to pages_with_children

File: cms/menu.py (hunk 3 above)
Comment (kux) [inline cms/menu.py]: I mean, the code DOES work if pages is sorted in ascending order and children ALWAYS have higher pks than parents, but these assumptions might not always be true.

File: cms/menu.py [hunk 5]
```diff
@@ -28,104 +62,118 @@ def get_visible_pages(request, pages, site=None):
...
-                    restricted_pages[id].append(perm)
-    # anonymous 
+    access_for_page = [ACCESS_PAGE]
+    access_for_children = [ACCESS_CHILDREN, ACCESS_PAGE_AND_CHILDREN]
+    access_for_desc = [ACCESS_DESCENDANTS, ACCESS_PAGE_AND_DESCENDANTS]
+    perms_to_collect = access_for_page + access_for_children + access_for_desc
+
+    # we'll get permission's user, group and users that belog to group just
+    #       so we save time executing more queries
+    perms_with_data = Page.objects.filter(
```
Comment (kux) [inline cms/menu.py]: perms_with_data sounds strange. It sounds like permissions + some other data. What data? And if it's about permissions, then why is the query executed on Page.objects?

What you're doing makes perfect sense, but the name of the variable doesn't suggest that. I think something like `page_ids_and_permissions` would be a more intuitive name.

File: cms/menu.py [hunk 6]
```diff
@@ -28,104 +62,118 @@ def get_visible_pages(request, pages, site=None):
...
+            'pagepermission__group__user').filter(
+                pagepermission__can_view=True,
+                pagepermission__grant_on__in=perms_to_collect).values_list(
+                    'id', 'pagepermission__id', 'pagepermission__grant_on',
+                    'pagepermission__user_id', 'pagepermission__group_id',
+                    'pagepermission__group__user__id'
+                    )
+
+    # separate data from the values returned by the query
+    perm_data = defaultdict(dict)
```
Comment (kux) [inline cms/menu.py]: `perm_data` makes sense, but at the end of the day everything is data.

So why not just call it permissions?

File: cms/menu.py [hunk 7]
```diff
@@ -28,104 +62,118 @@ def get_visible_pages(request, pages, site=None):
...
+    for data in perms_with_data:
+        page_id, perm_id, grant_on, user_id, group_id, user_from_group_id = data
+        perm_data[perm_id].setdefault('grant_on', grant_on)
```
Comment (kux) [inline cms/menu.py]: You don't need `setdefault`. Just do

```
perm_data[perm_id]['grant_on'] = grant_on
```

I think that's more readable

File: cms/menu.py [hunk 8]
```diff
@@ -28,104 +62,118 @@ def get_visible_pages(request, pages, site=None):
...
+        perm_data[perm_id].setdefault('page_id', page_id)
+        group = perm_data[perm_id].setdefault('group_id', [])
+        group.append(user_from_group_id) if user_from_group_id else ''
+
+    access_for_page += [ACCESS_PAGE_AND_CHILDREN, ACCESS_PAGE_AND_DESCENDANTS]
+    restricted = defaultdict(list)
+    for perm_id, data in perm_data.items():
+        page_id, grant_on = data['page_id'], data['grant_on']
+        # won't append it multiple times for the descendant/child pages
+        if perm_id in restricted[page_id]:
```
Comment (kux) [inline cms/menu.py]: How can a perm_id already be in a page's restricted list?

File: cms/menu.py [hunk 9]
```diff
@@ -200,96 +249,80 @@ def page_to_node(page, home, cut):
...
+            order_by("tree_id", "lft"))
+
+        nodes, first, home_cut, home = [], True, False, None
+
         # cache view perms
         visible_pages = get_visible_pages(request, pages, site)
+        titles = get_title_queryset(request).filter(
+            page__in=visible_pages,
+            language__in=fallbacks + [lang])
+        id_to_titles = defaultdict(dict)
```
Comment (kux) [inline cms/menu.py]: It's probably a good idea to rename this to `page_id_to_titles`

**Pull request merged**